### DISCUSSION TITLE
How to use test classes ?

### DISCUSSION THREAD

**Discussion by FSangouard** [eclipse-milo/milo#1237] [General]
Hello,

I am trying to add some unit tests to my project and when trying to connect to a test server I create before running my test suites, I get org.eclipse.milo.opcua.sdk.client.SessionFsm - [2] CreateSession failed: UaServiceFaultException: status=Bad_ServiceUnsupported, message=The server does not support the requested service.

I tried to mimic [ClientServerTest.java](https://github.com/eclipse/milo/blob/cd0f3d97f78de3823541229032642f8af9f00040/opc-ua-stack/stack-tests/src/test/java/org/eclipse/milo/opcua/stack/ClientServerTest.java) but either I missed something or I misunderstood the way these test classes should work.

From what I've seen in the stack test class, the test server setup does not include adding a SessionServiceSet, so I don't understand why mine should. Am I completely off the mark here?

**Comment by kevinherron** [comment]
It looks like this test sort of bypasses all that, and is mostly just trying to exercise the stack. At some point it sets a read handler that seems to be enough for it to send read requests/responses back and forth: https://github.com/eclipse/milo/blob/cd0f3d97f78de3823541229032642f8af9f00040/opc-ua-stack/stack-tests/src/test/java/org/eclipse/milo/opcua/stack/ClientServerTest.java#L151-L171

It must not have been that useful because it seems I've deleted it in `dev/1.0` and am relying more on the "full" integration tests using https://github.com/eclipse/milo/blob/master/opc-ua-sdk/integration-tests/src/test/java/org/eclipse/milo/opcua/sdk/test/AbstractClientServerTest.java

**Comment by kevinherron** [comment]
Ah, right, because in 1.0 there are no more "stack" client/server implementations. Just the full fat SDK client/server, and _transport_ implementations.

**Comment by FSangouard** [comment]
Thank you for your quick reply!

I had a look at AbstractClientServerTest.java, if my understanding is correct, all I need to do is make my test class extend this class and I should be able to connect to a local test server from within my test cases ?

**Reply by kevinherron** [reply]
A local instance of the Milo test server that it instantiates, yes. I'm not sure if a test artifact is published though, so you might have trouble extending this from your tests. This is an internal part of the integration tests, not something meant for public use.

**Reply by FSangouard** [reply]
Yeah, I tried adding artifacts sdk-tests and integration-tests as Maven dependencies, but I still can't import the class in my test class. Maven didn't download any Java classes.

So I guess if I want a local test server I have two options : build an artifact from the source or implement my own test server, right?

**Reply by kevinherron** [reply]
Yes that sounds about right. If you need to control the contents of the server it's best to build your own.
